src/main/scala/wdl4s/WdlNamespace.scala

```diff
@@ -143,76 +146,78 @@ case class WdlNamespaceWithWorkflow(importedAs: Option[String],
+
+  def apply(ast: Ast, source: WdlSource, importResolvers: Seq[ImportResolver], namespaceName: Option[String], root: Boolean = false): WdlNamespace = {
+
     val imports = for {
       importAst <- Option(ast).map(_.getAttribute("imports")).toSeq
       importStatement <- importAst.astListAsVector.map(Import(_))
     } yield importStatement
 
-    /**
-      * For import statements with an 'as' clause, this translates to a sub-namespace
-      */
-    val namespaces: Seq[WdlNamespace] = for {
-      imp <- imports if imp.namespaceName.isDefined
-      source = importResolver(imp.uri)
-      if source.length > 0
-    } yield WdlNamespace.load(source, imp.uri, importResolver, imp.namespaceName)
+    def tryResolve(str: String): Option[WdlSource] = {
+      importResolvers map { resolver =>
+        Try(resolver(str))
+      } collectFirst {
+        case Success(s) => s
+        case Failure(error) => throw error
+      }
+    }
 
     /**
-      * For import statements without an 'as' clause, this is treated as a c-style #include
+      * Translates all import statements to sub-namespaces
       */
-    val nonNamespacedImports = for {
-      imp <- imports if imp.namespaceName.isEmpty
-      source = importResolver(imp.uri)
-      ast = AstTools.getAst(source, imp.uri)
-      scopeAst <- ast.getAttribute("body").astListAsVector.collect({ case a: Ast => a })
-    } yield ast -> source
-
-    val topLevelScopeAsts = for {
-      namespaceAst <- nonNamespacedImports.map(_._1) :+ ast
-      scopeAst <- namespaceAst.getAttribute("body").astListAsVector.collect({ case a: Ast => a })
-    } yield scopeAst
+    val namespaces: Seq[WdlNamespace] = for {
+      imp <- imports
+      source = tryResolve(imp.uri)
+      if source.isDefined
+    } yield WdlNamespace.load(source.get, imp.uri, importResolvers, Option(imp.namespaceName))
 
     /**
       * Map of Terminal -> WDL Source Code so the syntax error formatter can show line numbers
       */
-    val nonNamespacedImportsTerminalMap = nonNamespacedImports.flatMap({ case (a, s) => AstTools.terminalMap(a, s) }).toMap
-    val terminalMap = nonNamespacedImportsTerminalMap ++ AstTools.terminalMap(ast, source)
+    val terminalMap = AstTools.terminalMap(ast, source)
     val combinedTerminalMap = ((namespaces map {x => x.terminalMap}) ++ Seq(terminalMap)) reduce (_ ++ _)
     val wdlSyntaxErrorFormatter = new WdlSyntaxErrorFormatter(combinedTerminalMap)
 
+    // Would like to rename this to something more appropriate
+    val topLevelAsts = ast.getAttribute("body").astListAsVector.collect({ case a: Ast => a })
+
     /**
-      * All `task` definitions, including ones from import statements with no 'as' clause
+      * All `task` definitions of primary workflow.
       */
-    val tasks: Seq[Task] = for {
-      taskAst <- topLevelScopeAsts if taskAst.getName == AstNodeName.Task
+    val topLevelTasks: Seq[Task] = for {
```

Comment:
Imports are still recursive, I believe I need this for the separation of tasks in the primary workflow vs imported workflows. In order to find a referenced task in a workflow with imports, the findTask() method uses the primary workflow tasks (topLevelTasks) when searching for calls without the "." Its possible that it can be restructured with less code?